src/chai/ArrayManager.inl

```diff
@@ -47,20 +47,26 @@ void* ArrayManager::reallocate(void* pointer, size_t elems, PointerRecord* point
   for (int space = CPU; space < NUM_EXECUTION_SPACES; ++space) {
     void* old_ptr = pointer_record->m_pointers[space];
 
-    if (old_ptr) {
-      pointer_record->m_user_callback(ACTION_ALLOC, ExecutionSpace(space), sizeof(T) * elems);
-      void* new_ptr = m_allocators[space]->allocate(sizeof(T)*elems);
+    pointer_record->m_user_callback(ACTION_ALLOC, ExecutionSpace(space), sizeof(T) * elems);
+    void* new_ptr = m_allocators[space]->allocate(sizeof(T)*elems);
```

Comment:
looks good

Reply:
This will cause reallocate to allocate a pointer in every space - I'm not sure that's what we want.

Reply:
@wrtobin  @robinson96  - I think this PR introduces an unexpected change in behavior, can you take a look?

Reply:
ooh - you're right David. 
Bill, the model should  preserve lazy allocation of device memory so that reallocations don't inadvertently initialize device memory if they haven't used gpu memory before.  At one point, we had an implementation of reallocate that also strode to only do a deep copy of the data on any memory space where the data was active.